File: src/minecraft/me/skrilled/utils/config/ConfigManager.java
package me.skrilled.utils.config;


import me.skrilled.SenseHeader;
import me.skrilled.utils.config.impl.ConfigFriends;
import me.skrilled.utils.config.impl.ConfigModule;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.LinkedList;

public class ConfigManager {

    private final LinkedList<AbstractConfig> abstractConfigs = new LinkedList<>();
    private final Path directory = Paths.get(String.valueOf(SenseHeader.getSense.getDirectory()),"configs");

    public ConfigManager() {
        this.getConfigs().clear();
        this.addConfigs(new ConfigModule(),new ConfigFriends());
        if(!this.directory.toFile().exists()) this.directory.toFile().mkdirs();
    }

    public void addConfigs(AbstractConfig... abstractConfigs) {
        this.getConfigs().addAll(Arrays.asList(abstractConfigs));
    }

    public void load(AbstractConfig abstractConfig) {
        if(abstractConfig != null && abstractConfig.getPath().toFile().exists()) abstractConfig.load();
    }

    public void loadAll() {
        this.getConfigs().stream().filter(abstractConfig -> abstractConfig.getPath().toFile().exists()).forEach(AbstractConfig::load);
    }

    public void save(AbstractConfig abstractConfig) {
        if(abstractConfig != null) abstractConfig.save();
    }

    public void saveAll() {
        this.getConfigs().forEach(abstractConfig -> {
            if(!abstractConfig.getPath().toFile().exists()) {
                try {
                    Files.createDirectories(abstractConfig.getPath().getParent());
                    abstractConfig.getPath().toFile().createNewFile();
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            abstractConfig.save();
        });
    }

    public LinkedList<AbstractConfig> getConfigs() {
        return abstractConfigs;
    }

    public Path getDirectory() {
        return directory;
    }
}

File: minecraft/me/kl0udy92/apart/config/ConfigManager.kt
package me.kl0udy92.apart.config

import me.kl0udy92.apart.core.Main
import org.reflections.Reflections
import java.nio.file.Files
import java.nio.file.Paths

class ConfigManager {

    val configs = mutableListOf<AbstractConfig>()
    private val directory = Paths.get(Main.directory.toString(), "configs")

    init {
        this.configs.clear()
        if (!this.directory.toFile().exists()) this.directory.toFile().mkdirs()
        Reflections("me.kl0udy92.${Main.name.lowercase()}.config.implementations").getSubTypesOf(AbstractConfig::class.java)
            .forEach {
                runCatching {
                    this.configs.add(
                        it.getDeclaredConstructor().newInstance()
                    )
                }.onFailure { exception -> exception.printStackTrace() }
            }
    }

    fun read(abstractConfig: AbstractConfig) {
        if (abstractConfig.path.toFile().exists()) abstractConfig.read()
    }

    fun readAll() {
        this.configs.filter { it.path.toFile().exists() }.forEach { it.read() }
    }

    fun write(abstractConfig: AbstractConfig) {
        if (abstractConfig.path.toFile().exists()) abstractConfig.write()
    }

    fun writeAll() {
        this.configs.forEach {
            kotlin.runCatching {
                Files.createDirectories(it.path.parent)
                it.path.toFile().createNewFile()
            }.onFailure { exception -> exception.printStackTrace() }
            it.write()
        }
    }

}